src/OpenTelemetry.Exporter.Zipkin/Implementation/ZipkinActivityConversionExtensions.cs

```diff
@@ -0,0 +1,219 @@
+﻿// <copyright file="ZipkinActivityConversionExtensions.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using OpenTelemetry.Internal;
+using OpenTelemetry.Resources;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Exporter.Zipkin.Implementation
+{
+    internal static class ZipkinActivityConversionExtensions
+    {
+        private static readonly Dictionary<string, int> RemoteEndpointServiceNameKeyResolutionDictionary = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase)
+        {
+            [SpanAttributeConstants.PeerServiceKey] = 0, // RemoteEndpoint.ServiceName primary.
+            ["net.peer.name"] = 1, // RemoteEndpoint.ServiceName first alternative.
+            ["peer.hostname"] = 2, // RemoteEndpoint.ServiceName second alternative.
+            ["peer.address"] = 2, // RemoteEndpoint.ServiceName second alternative.
+            ["http.host"] = 3, // RemoteEndpoint.ServiceName for Http.
+            ["db.instance"] = 4, // RemoteEndpoint.ServiceName for Redis.
+        };
+
+        private static readonly ConcurrentDictionary<string, ZipkinEndpoint> LocalEndpointCache = new ConcurrentDictionary<string, ZipkinEndpoint>();
+        private static readonly ConcurrentDictionary<string, ZipkinEndpoint> RemoteEndpointCache = new ConcurrentDictionary<string, ZipkinEndpoint>();
+
+        private static readonly DictionaryEnumerator<string, string, AttributeEnumerationState>.ForEachDelegate ProcessTagsRef = ProcessTags;
+        private static readonly ListEnumerator<ActivityEvent, PooledList<ZipkinAnnotation>>.ForEachDelegate ProcessActivityEventsRef = ProcessActivityEvents;
+
+        internal static ZipkinSpan ToZipkinSpan(this Activity activity, ZipkinEndpoint defaultLocalEndpoint, bool useShortTraceIds = false)
+        {
+            var context = activity.Context;
+            var startTimestamp = activity.StartTimeUtc.ToEpochMicroseconds();
+
+            string parentId = null;
+            if (activity.ParentSpanId != default)
```

Comment:
Might need to see the perf hit, this would translate to the following IL:

```
ldloc.0
callvirt instance valuetype [System.Diagnostics.DiagnosticSource]System.Diagnostics.ActivitySpanId [System.Diagnostics.DiagnosticSource]System.Diagnostics.Activity::get_ParentSpanId()
ldloca.s 1
initobj [System.Diagnostics.DiagnosticSource]System.Diagnostics.ActivitySpanId
ldloc.1
call bool [System.Diagnostics.DiagnosticSource]System.Diagnostics.ActivitySpanId::op_Equality(valuetype [System.Diagnostics.DiagnosticSource]System.Diagnostics.ActivitySpanId, valuetype [System.Diagnostics.DiagnosticSource]System.Diagnostics.ActivitySpanId)
```

Reply:
Great catch. Fixed creating object everytime.